solidity/ecdsa/deploy/17_upgrade_wallet_registry_v2.ts

```diff
@@ -0,0 +1,316 @@
+        return false
+      }
+    }
+  } catch (error) {
+    // V1 doesn't have allowlist() function - upgrade is needed
+    console.log("  Allowlist: not found (V1 detected - upgrade needed)")
+  }
+
+  // Get ProxyAdmin address from EIP-1967 slot
+  const ADMIN_SLOT = "0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103"
+  const proxyAdminSlot = await ethers.provider.getStorageAt(
+    walletRegistryDeployment.address,
+    ADMIN_SLOT
+  )
+  const proxyAdminAddress = ethers.utils.getAddress("0x" + proxyAdminSlot.slice(-40))
+  console.log(`  ProxyAdmin: ${proxyAdminAddress}`)
+
+  // Get current implementation
+  const IMPL_SLOT = "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"
+  const implSlot = await ethers.provider.getStorageAt(
+    walletRegistryDeployment.address,
+    IMPL_SLOT
+  )
+  const currentImpl = ethers.utils.getAddress("0x" + implSlot.slice(-40))
+  console.log(`  Current Implementation: ${currentImpl}`)
+
+  // Verify governance state is preserved
+  const governanceBefore = await walletRegistryBefore.governance()
+  console.log(`  Governance: ${governanceBefore}`)
+  console.log()
+
+  // Deploy new implementation
+  console.log("=== DEPLOYING NEW IMPLEMENTATION ===")
+  console.log()
+
+  const WalletRegistryFactory = await ethers.getContractFactory("WalletRegistry", {
+    signer: deployer,
+    libraries: {
+      EcdsaInactivity: EcdsaInactivity.address,
+    },
+  })
+
+  // Deploy implementation with constructor args (immutable variables)
+  const newImplementation = await WalletRegistryFactory.deploy(
+    EcdsaSortitionPool.address,
+    TokenStaking.address
+  )
+  await newImplementation.deployed()
+
+  console.log(`New implementation deployed: ${newImplementation.address}`)
+  console.log(`  TX: ${newImplementation.deployTransaction.hash}`)
+  console.log()
+
+  // Save deployment artifact
+  await deployments.save("WalletRegistryV2Implementation", {
+    address: newImplementation.address,
+    abi: WalletRegistryFactory.interface.format("json") as any,
+    transactionHash: newImplementation.deployTransaction.hash,
+  })
+
+  // Encode initializeV2 call
+  const initializeV2Data = WalletRegistryFactory.interface.encodeFunctionData(
+    "initializeV2",
+    [Allowlist.address]
+  )
+
+  // Encode upgradeAndCall for ProxyAdmin
+  const proxyAdminInterface = new ethers.utils.Interface([
+    "function upgradeAndCall(address proxy, address implementation, bytes calldata data) external payable",
+    "function owner() external view returns (address)"
+  ])
+
+  const upgradeCalldata = proxyAdminInterface.encodeFunctionData(
+    "upgradeAndCall",
+    [walletRegistryDeployment.address, newImplementation.address, initializeV2Data]
+  )
+
+  // Get ProxyAdmin owner
+  const proxyAdmin = await ethers.getContractAt(
+    ["function owner() view returns (address)"],
+    proxyAdminAddress
+  )
+  const proxyAdminOwner = await proxyAdmin.owner()
```

Comment:
maybe for mainnet deployments, consider adding a sanity check that `proxyAdminOwner` is actually the timelock contract address, not an EOA